[SOW 3] Fix race conditions in job startup

njsmith opened the pull request:
This isn't all of SOW 3 yet, but since it's a coherent piece I figured
I'd post it, and you can decide whether you want to review as you go
or wait for the rest. (Next step is docker support.)


The flask server no longer writes to run_dir directly during job startup;
instead, it writes to a temporary directory, and then passes ownership of
the temporary directory to the runner daemon. This is the first step
towards the flask server never touching the run_dir directly.

Under normal circumstances, the temp dir is always cleaned up immediately
after creation. In case of abnormal circumstances, the runner daemon now
checks once a day for any stale temp dirs, and cleans them up.

Also, the runner daemon is now careful to make sure that running jobs
always correspond to the run_dir on disk -- stale jobs are killed
before setting up a new run_dir, etc.


Previously, we discussed having the runner daemon pick a temporary
name and pass it to server.py, with the idea being that this would
help the runner daemon clean up any garbage. But I changed it -- under
normal circumstances we never leak garbage, so the only reason to gc
is if something weird happened (e.g. process crash at an unfortunate
time), and "something weird" could just as easily happen to the runner
daemon as it could to server.py. So instead we use a stupid simple
rule: any directory named "*.tmp" whose mtime is >24 hours ago, is
garbage. This lets us reliably clean up garbage regardless of how it
got there, and means server.py can pick its own temp dirs.

QUESTION: should this be connected to simulation_db.tmp_dir() somehow?
From a quick skim that seems to be used in a bunch of places without
any cleanup at all, so I wasn't sure what the intended semantics
were and just did my own thing...

robnagler commented:
Keep going. I'll review later. Sounds all good.

We'll deal with simulation_db.tmp_dir later.

njsmith commented:
This is now feature complete WRT the items in SOW-3.

Specifically, it includes:

* All known race conditions around job lifecycles are fixed
* There's a docker backend
* Backend can be selected on a per-job basis. (Currently `server.py` always uses `local`, unless `SIREPO_FEATURE_CONFIG_RUNNER_DAEMON_DOCKER=1` is set, in which case it always uses `docker`.)
* The docker backend doesn't run jobs directly in `run_dir`; instead it creates a `working_dir`, and runs jobs there. So if `server.py` tries to read results out of `run_dir` directly, it will fail.
* We now have two kinds of jobs: "report jobs", and "extract jobs". Report jobs are what we had before: each run has a unique hash, they run in the background, you can poll for their status, etc. Extract jobs are new, and are what `server.py` now uses to fetch the results of report jobs.
* The following operations now use extract jobs instead of poking around in the `run_dir` directly:
  * `remove_last_frame` (run after a report is cancelled)
  * `background_percent_complete`
  * `get_simulation_frame`
  * general result extraction after a run completes
* The runner records which backend was used to run each report job, and then automatically uses the same backend for any associated extract jobs. Also, the backend can pass some arbitrary data from the report job to the extract job (e.g., the working directory, or in the future: which docker host was selected to hold the working directory, etc.)

There are likely a few bugs somewhere that I didn't catch yet :-). But I have confirmed that we can run myapp, warppba, and warpvnd, both the short runs that the GUI does automatically, and the long `Start new simulation` run with a progress bar + incremental video frames.

General theory:

* Every report still has a `run_dir`, which still holds the input parameters, plus status information (controlled by the runner). This lets us avoid having to immediately rewrite all the code in `server.py` that directly reads `in.json`.
* Actual jobs happen in a "working dir". In the local process backend, the working dir = the run dir. In the docker backend, they're different.... from the host's perspective. Inside the container, we mount the working dir at the location where the code is expecting to find the run dir. Again, this lets us avoid having to rewrite a bunch of code that assumes it knows the run dir's hard-coded path.
* To run an extract job, you do `runner_client.run_extract_job(run_dir, jhash, type, *args)`. This is a blocking operation – it waits for the job to complete before returning. The extract job holds the `run_dir` lock while running, so extract jobs inside each `run_dir` happen in serial order.
* Extract jobs eventually run `sirepo extract <type> <args>`, where `<args>` are arbitrary JSON-encoded values. And this writes out some arbitrary JSON to stdout, which eventually becomes the return value from `runner_client.run_extract_job`. The runner daemon+backend are responsible for proxying this data around.

This means that the docker backend needs to be able to run `sirepo extract` *inside the container*. To test this, therefore, you need a container image with *this branch* installed in it. To test, I used a hack like:

```diff
modified   container-conf/build.sh
@@ -29,8 +29,9 @@ build_as_run_user() {
     cd "$build_guest_conf"
     sirepo_boot_init
     git clone -q --depth=50 https://github.com/radiasoft/pykern
-    git clone -q --depth=50 "--branch=${TRAVIS_BRANCH:-master}" \
-        https://github.com/radiasoft/sirepo
+    # git clone -q --depth=50 "--branch=${TRAVIS_BRANCH:-master}" \
+    #     https://github.com/radiasoft/sirepo
+    git clone -q --depth=50 --branch=njs-sow-3 https://github.com/njsmith/sirepo
     cd sirepo
     if [[ ${TRAVIS_COMMIT:+1} ]]; then
         git checkout -qf "$TRAVIS_COMMIT"
modified   sirepo/runner_daemon/docker_process.py
@@ -77,7 +77,8 @@ async def _make_container(run_dir, working_dir, quoted_bash_cmd, job_type):
     return await trio.run_sync_in_worker_thread(
         functools.partial(
             _DOCKER.containers.run,
-            'radiasoft/sirepo',
+            #'radiasoft/sirepo',
+            'docker.io/radiasoft/sirepo:20190319.073907',
             docker_cmd,
             # Mount the host's working_dir at run_dir inside the container
             # {host path: {'bind': container path, 'mode': 'rw'}}
```

This is a bit awkward, though hopefully the `extract` commands won't change often...?

When testing, I did notice that the full 56-frame animation run was a bit awkward. Even after the job finished, when I tried playing the animation, it seemed to be going back to the server and running a new extract job for each frame it played. It turns out that setting up and tearing down a docker image on every frame doesn't do wonders for your frame rate. I guess we'll probably want to figure out better caching? But let's get it working before we worry about making it fast. (I'm not sure why the browser wasn't caching it.)

robnagler commented on sirepo/pkcli/runner.py:
```diff
@@ -15,31 +15,33 @@
+# Every _TMP_DIR_CLEANUP_TIME seconds, we scan through the run database, and
+# any directories that are named '*.tmp', and whose mtime is
+# >_TMP_DIR_CLEANUP_TIME in the past, are deleted.
+_TMP_DIR_CLEANUP_TIME = 24 * 60 * 60  # 24 hours
```
No comment needed. Use Napoleon/Sphinx/Google `#:` prefix to comments (above) for doc.

njsmith commented on sirepo/pkcli/extract.py:
```diff
@@ -0,0 +1,72 @@
+    )
+
+
+# XX FIXME: can we avoid passing the whole data dict through here?
```
The main issue is the `'models' in data` check, which IIUC is actually a proxy for checking whether this is the first time we're extracting the result from a given `run_dir`, because somehow the distributed state machine managed by javascript+flask makes that work as a way to check. I think this function really only needs a few bits of `data`, and if we sat down and traced it out then it would be pretty straightforward to decouple it from the js/flask details. I just didn't trust myself to untangle that at the same time I was waist-deep in implementing the extract job machinery.

robnagler replied:
I think we'll have to wait on this optimization. See #1557 for discussion

njsmith commented on sirepo/runner_daemon/docker_process.py:
```diff
@@ -0,0 +1,161 @@
+    # through bash to set up our environment. But we do assume that it's
+    # sufficiently well-behaved that shlex.quote will work.
+    docker_cmd = [
+        '/bin/bash',
```
> I wonder if we should't be using `bash -l -c cmd`.

Eh, I guess it works the same either way, and `-l` is a bit simpler than `. .bashrc &&`, so sure, why not.

> I think we'll want jobs to be run from a template bash file that passes arguments through a file. This connects to the args serialization comments in runner_client and extract. We always need some sort of work_dir.

A template seems like overkill? The code to massage the command into the appropriate form is only a few lines, and if you don't use files there's no risk of file collisions. But it's your call of course :-)

robnagler replied:
We will need a wrapper inside the container to handle some state management, e.g. background percent reporting or simply "i'm alive". With a template, we document what parameters need to be passed, e.g. SIREPO_MPI_CORES and PYKERN_PKCONFIG_CHANNEL (not being passed, but should be). 

Let's discuss for SOW-4. I'll create an issue.

njsmith commented on sirepo/runner_daemon/docker_process.py:
```diff
@@ -0,0 +1,161 @@
+            docker_cmd,
+            # Mount the host's working_dir at run_dir inside the container
+            # {host path: {'bind': container path, 'mode': 'rw'}}
+            volumes={str(working_dir): {'bind': str(run_dir), 'mode': 'rw'}},
```
Yeah, it would be nice to make development work without having to regenerate the docker container. Mounting `~/.pyenv` is a neat idea, but I think it's more complicated than that, because for development we use in-place installs?

njsmith commented on sirepo/runner_daemon/docker_process.py:
```diff
@@ -0,0 +1,161 @@
+
+# Helper to call container.wait in a async/cancellation-friendly way
+async def _container_wait(container):
+    while True:
```
Then `container.wait` raises `docker.errors.NotFound`, and in `_supervise_report_job` the `finally` block sets the job status to `ERROR`, and then the `NotFound` error and traceback are logged by `_catch_and_log_errors`.

njsmith commented on sirepo/runner_daemon/docker_process.py:
```diff
@@ -0,0 +1,161 @@
+        pass
+    else:
+        pkdlog('found stale container {}; removing', name)
+        await trio.run_sync_in_worker_thread(old_container.remove)
```
Not for any good reason :-). And if we genuinely fail to remove a container with the name we want to use, then the next step is going to fail regardless. But looking at it again, I guess it would actually make the code a bit simpler to allow `NotFound` from both `containers.get` or `container.remove`, and that does remove a theoretical race condition that we'd never hit in practice.

njsmith commented on sirepo/pkcli/runner.py:
```diff
@@ -105,122 +93,192 @@ def _write_status(status, run_dir):
-                # directly into the run_dir.
+            disk_status = pkio.read_text(disk_status_path)
+            if disk_status == 'pending':
+                # We never write this, so it must be stale, in which case
```
In `run_dir_status` we don't know if the user wants to run the job again or what. But in `api_runSimulation`, if the current status is `ERROR` then we do re-start the job.

We use the status file to remember the state of already-finished jobs, that the runner isn't otherwise tracking anymore. By the time we save the `disk_status`, the `run_dir` is basically immutable, so there's no point in keeping in-memory data structures. Technically we don't need to do it this way, but we need some way to spill the status to disk to get it out of RAM, and this is good as anything – plus I'm pretty sure there's still code in `simulation_db` that expects the status file to be there.

njsmith commented on sirepo/pkcli/runner.py:
```diff
@@ -15,31 +15,33 @@
-# need to pass any of that on, just like runner.docker, doesn't
-_EXEC_ENV_REMOVE = re.compile('^(OMPI_|PMIX_|SIREPO_|PYKERN_)')
+
+_BACKEND_INFO_BASENAME = 'backend.json'
```
We save the backend info before starting to run the job, and we write the status after it's done. I suppose we could use the same file and rewrite it, but the lifecycle and consumers are otherwise unrelated, so using two files keeps it decoupled. It doesn't matter too much either way I guess.

njsmith commented on sirepo/pkcli/runner.py:
```diff
@@ -105,122 +93,192 @@ def _write_status(status, run_dir):
+        tmp_dir.rename(run_dir)
+        # Start the job:
+        report_job = await _BACKENDS[backend].start_report_job(run_dir, cmd)
+        # And update our records so we know it's running:
```
The locking strategy is: each `run_dir` has an associated lock. Everything that touches the `run_dir`-related data structures holds that lock. Currently this is done mostly by `handle_conn` (it takes the lock before calling the handler), and then `_supervise_report_job` also takes the lock after the job finishes so it can write back the new state.

I'm not very happy about how implicit this is in the code, and have been trying to figure out if there's a way to make it more clear. (Maybe using reentrant locks and re-acquiring the lock in every method?)

njsmith commented on sirepo/pkcli/runner.py:
```diff
@@ -281,13 +342,33 @@ def _rpc_handler(fn):
+    count = 0
+    cutoff = time.time() - _TMP_DIR_CLEANUP_TIME
+    for dirpath, dirnames, filenames in os.walk(srdb.root()):
+        if (dirpath.endswith('.tmp')
```
Good point. Since it's a global constant related to the sirepo db and used in multiple files, I moved it into `sirepo.srdb`.

njsmith commented on sirepo/runner_daemon/docker_process.py:
```diff
@@ -0,0 +1,161 @@
+
+async def run_extract_job(run_dir, cmd, backend_info):
+    # no output redirection here - we want to let the docker daemon collect it
+    quoted_bash_cmd = ' '.join(shlex.quote(c) for c in cmd)
```
The extract jobs are designed to do all their communication without using the disk, to avoid a chicken-and-egg problem: if we save the extract job output into the working dir, then we need to run another extract job to get the results from the first extract job! Also, if there are multiple extract jobs, you have to give them unique names and then keep track of which is which. Given that we already need a mechanism to ship back the return code and stdout, I think it's simplest to ship back stderr as well, and then dump it into the log in the sirepo server.

njsmith commented on sirepo/runner_daemon/local_process.py:
```diff
@@ -0,0 +1,114 @@
+
+class _LocalReportJob:
+    # No backend-specific per-process info
+    backend_info = {}
```
sure

njsmith commented on sirepo/pkcli/runner.py:
```diff
@@ -105,122 +93,192 @@ def _write_status(status, run_dir):
+            finally:
+                async with self.locks[run_dir]:
+                    # Clear up our in-memory status
+                    assert self.report_jobs[run_dir] is job_info
```
Eh, I guess `is` is probably a tiny smidgen faster. But really I used `is` because conceptually, what we want to check here is that no-one's replaced the `job_info` object behind our back, no matter how similar it might or might not be. In this case it makes no difference, because `==` on `_JobInfo` uses `object.__eq__` which does an `is` comparison anyway, but if we did have an `__eq__` override then I think it would still be correct to use `is` here.

njsmith commented on sirepo/runner_client.py:
```diff
@@ -45,23 +45,35 @@ def _rpc(request):
+        'run_dir': str(run_dir),
+        'jhash': jhash,
+        'subcmd': subcmd,
+        'args': [pkjson.dump_pretty(arg) for arg in args],
```
Good idea!

njsmith commented on sirepo/pkcli/runner.py:
```diff
@@ -105,122 +93,192 @@ def _write_status(status, run_dir):
+            pkdlog(
+                '{} {}: got output on stderr:\n{}',
+                run_dir, jhash,
+                result.stderr.decode('utf-8', errors='ignore'),
```
The `utf-8, errors='ignore'` is because the output should always be valid utf-8, but if something goes terribly wrong, it's better to get mangled debug output than to suffer a double-fault.

The `latin1` is because I originally forgot how to spell `utf-8, errors='ignore'`, and then after I remembered, I missed updating one of the places where I'd written it the wrong way. That's fixed now.

For writing the output, we can't make it intermingled, because `stdout` is where the structured data is. (And IIRC some of the extract jobs dump random commentary to stderr, because it's scientific code.) And see my comment above re: the chicken-and-egg problems of putting stderr into a file.

robnagler replied:
This is the issue encapsulation should solve:

> originally forgot how to spell `utf-8, errors='ignore'`

Perhaps we need something in pkdlog. In the past, I've also avoided massive log output by limiting nested data structure recursion depth and argument size. This is a similar problem: you don't want to write control characters or other junk to the logs. If you have examples of how people encapsulate utf8 and/or these other issues, I'd appreciate it.

Try to remember to put a TODO there at some point. BTW, we use [Google's TODO mechanism](http://google.github.io/styleguide/pyguide.html#312-todo-comments): `#TODO(rjn)` or `#TODO(robnagler)` so we don't have to blame to see who put the TODO there. We do not leave a space between the `#` and `TODO` to allow simple grepping. Not a big deal, just fix next SOW.

njsmith commented:
OK, I think I've addressed or responded to all the review comments.

njsmith commented on sirepo/pkcli/runner.py:
```diff
@@ -105,122 +93,192 @@ def _write_status(status, run_dir):
         disk_in_path = run_dir.join('in.json')
         disk_status_path = run_dir.join('status')
         if disk_in_path.exists() and disk_status_path.exists():
+            # status should be recorded on disk XOR in memory
```
If status exists in both places then it means that somehow our job startup or job teardown code didn't do its job properly, so there's probably more going wrong than just having two copies of the status. It's caught some bugs already.
